File: exercise/java/day12/src/main/java/gifts/Santa.java
package gifts;

import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;

public class Santa {

    private final List<Child> childrenRepository;
    public Santa() {
        this.childrenRepository = new ArrayList<>();
    }

    public Toy chooseToyForChild(String childName){
        Optional<Child> found = Optional.empty();
        for (int i = 0; i < childrenRepository.size(); i++) {
            Child currentChild = childrenRepository.get(i);
            if (currentChild.getName().equals(childName)) {
                found = Optional.of(currentChild);
            }
        }
        Child child = found.orElseThrow(NoSuchElementException::new);

        if("naughty".equals(child.getBehavior()))
            return child.getWishlist().get(child.getWishlist().size() - 1);

        if("nice".equals(child.getBehavior()))
            return child.getWishlist().get(1);

        if("very nice".equals(child.getBehavior()))
            return child.getWishlist().get(0);

        return null;
    }

    public void addChild(Child child) {
        childrenRepository.add(child);
    }
}

File: exercise/kotlin/day12/src/main/kotlin/gifts/Santa.kt
package gifts

class Santa {

    private val childrenRepository = mutableListOf<Child>()

    fun chooseToyForChild(childName: String): Toy {
        var found: Child?=null
        for (i in childrenRepository.indices) {
            val currentChild: Child = childrenRepository[i]
            if (currentChild.name == childName) {
                found = currentChild
            }
        }
        if(null == found){
            throw NoSuchElementException("Cannot ind child $childName")
        }
        val child: Child = found

        val wishlist = child.getWishlist()!!
        if ("naughty" == child.behavior) return wishlist[wishlist.size - 1]
        if ("nice" == child.behavior) return wishlist[1]
        return if ("very nice" == child.behavior) wishlist[0] else throw IllegalStateException("unknown child behavior $child")
    }

    fun addChild(child: Child) {
        childrenRepository.add(child)
    }
}
